### PR TITLE
Issue 6340: Fix CommitRequestHandler may get stuck throwing NPE

### LINKED ISSUE #6340: CommitRequestHandler may get stuck throwing NPE
**Describe the bug**
Some debug on one of the failed system tests we had recently:

- The Controller starts working on `testReadTxnWriteAutoScaleStream` and few seconds later starts trying to commit transactions:
```
2021-09-22 10:33:51,710 142230 [eventprocessor-49] DEBUG i.p.c.s.e.r.CommitRequestHandler - [requestId=-4778385706480942802] Attempting to commit available transactions on stream testReadTxnWriteAutoScaleScope2024571197/testReadTxnWriteAutoScaleStream
```
- As part of that process, there are some table reads that need to be issued to resume committing transactions for this new controller:
```
2021-09-22 10:33:51,719 142239 [ClientSocketReaders-8] DEBUG i.p.controller.server.SegmentHelper - [requestId=-4778385706480942802] ReadTable _system/_tables/testReadTxnWriteAutoScaleScope2024571197/testReadTxnWriteAutoScaleStream/metadata.#.eefe0628-bdc8-427d-9aa6-9719279e95c2 TableRead 1116691496960.
2021-09-22 10:33:51,722 142242 [controllerpool-19] DEBUG i.p.c.store.host.ZKHostStore - Found owning host: 172.26.48.21:12345: for containerId: 0
...
2021-09-22 10:33:51,732 142252 [ClientSocketReaders-1] DEBUG i.p.controller.server.SegmentHelper - [requestId=-4778385706480942802] ReadTable _system/_tables/testReadTxnWriteAutoScaleScope2024571197/testReadTxnWriteAutoScaleStream/transactionsInEpoch-1.#.eefe0628-bdc8-427d-9aa6-9719279e95c2 TableRead 1125281431552.
```

- Reads seem to have completed correctly. However, the Controller has some problem when processing the data read from that table (`transactionsInEpoch`), as it seems to be either incomplete or directly null:
```
2021-09-22 10:33:51,732 142252 [ClientSocketReaders-1] DEBUG i.p.controller.server.SegmentHelper - [requestId=-4778385706480942802] ReadTable _system/_tables/testReadTxnWriteAutoScaleScope2024571197/testReadTxnWriteAutoScaleStream/transactionsInEpoch-1.#.eefe0628-bdc8-427d-9aa6-9719279e95c2 TableRead 1125281431552.
2021-09-22 10:33:51,735 142255 [eventprocessor-41] WARN  i.p.c.s.e.r.CommitRequestHandler - [requestId=-4778385706480942802] Exception while attempting to commit transaction on stream testReadTxnWriteAutoScaleScope2024571197/testReadTxnWriteAutoScaleStream
java.util.concurrent.CompletionException: java.lang.NullPointerException
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.complete(Unknown Source)
	at io.pravega.controller.store.PravegaTablesStoreHelper.lambda$getEntries$45(PravegaTablesStoreHelper.java:557)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(Unknown Source)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at io.pravega.common.concurrent.ThreadPoolScheduledExecutorService$ScheduledRunnable.run(ThreadPoolScheduledExecutorService.java:186)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.NullPointerException: null
	at io.pravega.controller.store.stream.records.ActiveTxnRecord.getCommitOrder(ActiveTxnRecord.java:93)
	at io.pravega.controller.store.stream.PravegaTablesStream.lambda$getVersionedTransactionRecords$86(PravegaTablesStream.java:912)
	... 12 common frames omitted
```
- In method `PravegaTablesStream:getVersionedTransactionRecords` (which has been recently added as part of PR #6283) there is no logic to expect a null active transaction record, so the problem is persistently there. Now, the question is: why the data stored in transactionsInEpoch is null or incomplete? If this cannot be accepted, maybe we need how this data has been written and if it is a valid case, we may need to do a better handling.

**To Reproduce**
Appears sporadically on system tests.

**Screenshots**
n/a

**Additional information**
n/a

### PR DESCRIPTION
**Pull request by pbelgundi**
Signed-off-by: pbelgundi <prajakta.belgundi@emc.com>

**Change log description**  
When pulling entries for transactions from the transactionsInEpoch table, if no entry is found for a particular transaction in this table, the WireCommand.ReadTable returns a TableSegmentEntry with key set to WireCommands.TableKey.NOT_EXISTS. On receiving this response on Controller, SegmentHelper adds an ActiveTransactionRecord.EMPTY as the transaction record for this particular transaction. This ActiveTransactionRecord.EMPTY record has the commitOrder and commitOffsets parameters set to null leading to NPE when such a record is returned.

**Purpose of the change**  
Fixes #6340 

**What the code does**  
Changed ActiveTxnRecord.EMPTY record so there are no data members with null values.
Changed code such that, If metadata for a certain txnId is not found in Segment Store, skip adding it to  List<VersionedTransactionData> instead of adding ActiveTxnRecord.EMPTY record.

**How to verify it**  
All tests should pass, especially system test - testReadTxnWriteAutoScaleStream

### CODE REVIEW

**Review by shrids — CHANGES_REQUESTED** [review]
Blocking it until we have tests to validate this fixes the NPE.

**Comment by pbelgundi** [comment]
System test run on PR branch completed successfully without any errors. Checked Controller logs for NPE exceptions and could not find any. 
Will do multiple system test runs to confirm this is fixed since the issue is a super corner case and may not be hit on most occasions.

File: controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStream.java [hunk 1]
```diff
@@ -905,13 +903,14 @@ private String getWritersTableName(String id) {
...
                                      VersionedMetadata<ActiveTxnRecord> txn = res.get(i);
                                      ActiveTxnRecord activeTxnRecord = txn.getObject();
-
-                                     VersionedTransactionData vdata = new VersionedTransactionData(epoch, UUID.fromString(txnIds.get(i)), txn.getVersion(),
-                                             activeTxnRecord.getTxnStatus(), activeTxnRecord.getTxCreationTimestamp(),
-                                             activeTxnRecord.getMaxExecutionExpiryTime(), activeTxnRecord.getWriterId(),
-                                             activeTxnRecord.getCommitTime(), activeTxnRecord.getCommitOrder(),
-                                             activeTxnRecord.getCommitOffsets());
-                                     list.add(vdata);
+                                     if (!ActiveTxnRecord.EMPTY.equals(activeTxnRecord)) {
```
Comment (shrids) [inline controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStream.java:906]: If the txnId does not exist in the `epochTxnTable` then we observe an `ActiveTxnRecord.EMPTY`. i.e. when the result is NON_EXISTENT_TXN.  Is this the only scenario where commit order is null?

Please fix this at the defnition of `io.pravega.controller.store.stream.records.ActiveTxnRecord#EMPTY` as this is the where the commitOrder should be inited to a non-null value. 
This will ensure the remaining methods in the PravegaTablesStream also need not have a check for ActiveTxnRecord.EMPTY.

**Comment by pbelgundi** [comment]
> Blocking it until we have tests to validate this fixes the NPE.

Added tests to cover the failing scenario.

File: controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStream.java (hunk 1 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStream.java:906]: I have already done so.

Comment (shrids) [reply controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStream.java:906]: Thanks for pointing it out.
Any specific reason why we still have this check  `if (!ActiveTxnRecord.EMPTY.equals(activeTxnRecord))` ?
If it is required then `getTransactionRecords` in line 878 also has a similar pattern.

File: controller/src/test/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessorPravegaTablesStreamTest.java [hunk 2]
```diff
@@ -15,18 +15,116 @@
...
+import io.pravega.controller.store.stream.records.CommittingTransactionsRecord;
+import io.pravega.controller.store.task.TaskStoreFactory;
+import io.pravega.controller.task.EventHelper;
+import io.pravega.controller.task.Stream.StreamMetadataTasks;
+import io.pravega.controller.task.Stream.StreamTransactionMetadataTasks;
+import io.pravega.controller.util.Config;
+import io.pravega.shared.controller.event.CommitEvent;
+import io.pravega.test.common.AssertExtensions;
+import org.junit.Test;
+import  static org.junit.Assert.assertFalse;
```
Comment (RaulGracia) [inline controller/src/test/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessorPravegaTablesStreamTest.java:47]: These 2 static imports have an extra initial space and an unrecognized character by Git (perhaps due to the editor used?)

File: controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStream.java [hunk 3]
```diff
@@ -991,7 +990,7 @@ private String getWritersTableName(String id) {
     @Override
     CompletableFuture<Void> removeTxnsFromCommitOrder(List<Long> orderedPositions, OperationContext context) {
         Preconditions.checkNotNull(context, "operation context cannot be null");
-
+        log.debug("Removing TRANSACTIONS FROM COMMIT ORDERER");
```
Comment (RaulGracia) [inline controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStream.java:993]: Recall to remove these debug messages.

File: controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStream.java [hunk 4]
```diff
@@ -216,7 +216,7 @@ private String getWritersTableName(String id) {
                                                                   OperationContext context,
                                                                   Map<String, TxnWriterMark> writerMarks) {
         Preconditions.checkNotNull(context, "operation context cannot be null");
-
+        log.debug("completeCommittingTransactions.");
```
Comment (shrids) [inline controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStream.java:219]: this log line can have more context.

File: controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStream.java (hunk 1 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStream.java:906]: I don;t see the point in doing this check in getTransactionRecords. The check `if (!ActiveTxnRecord.EMPTY.equals(activeTxnRecord))` is needed so we do not update the information for EMPTY ActiveTxnRecords when updating writer marks.

File: controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStream.java (hunk 4 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStream.java:219]: Didn;t intend to add this log , had added while debugging, got left out by mistake... will remove.

File: controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStream.java (hunk 3 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStream.java]: fixed

File: controller/src/test/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessorPravegaTablesStreamTest.java (hunk 2 above)
Comment (pbelgundi) [reply controller/src/test/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessorPravegaTablesStreamTest.java]: fixed

File: controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStream.java (hunk 4 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStream.java]: fixed

Comment (shrids) [reply controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStream.java]: ok

File: controller/src/test/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessorPravegaTablesStreamTest.java (hunk 2 above)
Comment (pbelgundi) [reply controller/src/test/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessorPravegaTablesStreamTest.java]: > System test run on PR branch completed successfully without any errors. Checked Controller logs for NPE exceptions and could not find any. Will do multiple system test runs to confirm this is fixed since the issue is a super corner case and may not be hit on most occasions.

So far completed 5 system test runs on the PR branch and none had any NullPointerExceptions in Controller logs.

**Review by shrids — APPROVED** [review]
No furhter comments from my end...

**Review by RaulGracia — APPROVED** [review]
